jena-arq/src/main/java/org/apache/jena/sparql/engine/join/ImmutableUniqueList.java

```diff
@@ -0,0 +1,208 @@
+         */
+        private Collection<T> items;
+
+        Builder(Class<T> itemClass) {
+            super();
+            this.itemClass = itemClass;
+        }
+
+        private void alloc(int n) {
+            if (items == null) {
+                items = n < INDEX_THRESHOLD ? new ArrayList<>(INDEX_THRESHOLD) : new LinkedHashSet<>();
+            } else if (!(items instanceof Set) && (items.size() + n >= INDEX_THRESHOLD)) {
+                Set<T> tmp = new LinkedHashSet<>(items);
+                items = tmp;
+            }
+        }
+
+        public Builder<T> add(T item) {
+            if (!(items instanceof Set)) {
+                if ( items == null || ! items.contains(item) ) {
+                    alloc(1);
+                    items.add(item) ;
+                }
+            } else {
+                items.add(item);
+            }
+            return this ;
+        }
+
+        public Builder<T> addAll(Collection<T> items) {
+            alloc(items.size());
+            for (T item : items) {
+                add(item);
+            }
+            return this;
+        }
+
+        public Builder<T> addAll(T[] arr) {
+            alloc(arr.length);
+            for (T item : arr) {
+                add(item);
+            }
+            return this;
+        }
+
+        public Builder<T> remove(Object o) {
+            if (items != null) {
+                items.remove(o) ;
+            }
+            return this ;
+        }
+
+        public Builder<T> clear() {
+            items = null;
+            return this ;
+        }
+
+        public boolean isEmpty() {
+            return items == null || items.isEmpty();
+        }
+
+        @SuppressWarnings("unchecked")
+        public ImmutableUniqueList<T> build() {
+            T[] finalItems;
+            if (items == null) {
+                finalItems = (T[])Array.newInstance(itemClass, 0);
+            } else {
+                finalItems = (T[])Array.newInstance(itemClass, items.size());
+                items.toArray(finalItems);
+            }
+            return new ImmutableUniqueList<>(INDEX_THRESHOLD, finalItems);
+        }
+    }
+
+    public static <T> Builder<T> newUniqueListBuilder(Class<T> itemClass) {
+        return new Builder<>(itemClass);
+    }
+
+    public static <T> ImmutableUniqueList<T> createUniqueList(Class<T> itemClass, Collection<T> items) {
+        return ImmutableUniqueList.<T>newUniqueListBuilder(itemClass).addAll(items).build();
+    }
+
+    public static <T> ImmutableUniqueList<T> createUniqueList(Class<T> itemClass, T[] items) {
+        return ImmutableUniqueList.<T>newUniqueListBuilder(itemClass).addAll(items).build();
+    }
+
+    /**
+     * Create a join key without coping the key array and without checking for duplicates.
+     * The array must not be modified.
+     */
+//    @SuppressWarnings("unchecked")
+//    public static <T> ImmutableUniqueList<T> createUniqueListUnsafe(int indexThreshold, T[] elementData) {
```

Comment:
Could this be removed if it is not used?

Reply:
Yes, should be removed.

Reply:
done.